JuliaLowering/src/ast.jl

```diff
@@ -126,48 +126,49 @@ function _append_nodeids!(graph::SyntaxGraph, ids::Vector{NodeId}, vals::SyntaxL
     append!(ids, vals.ids)
 end
 
-function makeleaf(graph::SyntaxGraph, srcref, proto; attrs...)
+function makeleaf(graph::SyntaxGraph, srcref, proto::Union{Kind, SyntaxTree})
     id = newnode!(graph)
     ex = SyntaxTree(graph, id)
     copy_attrs!(ex, proto, true)
-    setattr!(graph, id; source=_unpack_srcref(graph, srcref), attrs...)
+    ex.source = _unpack_srcref(graph, srcref)
     return ex
 end
 
-function _makenode(graph::SyntaxGraph, srcref, proto, children; attrs...)
-    id = newnode!(graph)
-    setchildren!(graph, id, children)
-    ex = SyntaxTree(graph, id)
-    copy_attrs!(ex, proto, true)
-    setattr!(graph, id; source=_unpack_srcref(graph, srcref), attrs...)
-    return SyntaxTree(graph, id)
-end
-function _makenode(ctx, srcref, proto, children; attrs...)
-    _makenode(syntax_graph(ctx), srcref, proto, children; attrs...)
+function makeleaf(ctx::AbstractLoweringContext, srcref, proto)
+    makeleaf(syntax_graph(ctx), srcref, proto)
 end
 
-function makenode(ctx, srcref, proto, children...; attrs...)
-    _makenode(ctx, srcref, proto, _node_ids(syntax_graph(ctx), children...); attrs...)
+function makeleaf(ctx, srcref, proto, @nospecialize(attrs::AbstractVector))
```

Comment:
Another normalization trick you might consider trying out at some point is:
```julia
@noinline function _makeleaf(..., attrs::Vector{Any})
    # ... (complicated function)
end
@inline function makeleaf(..., @nospecialize(attrs::AbstractVector))
    return _makeleaf(..., convert(Vector{Any}, attrs))
end
```

This preserves "good" codegen, while trying to limit specialized code to the caller side. Ofc, it's even better if we can avoid the `Vector` type explosion to begin with.